editor/test/integration/editor_extensions_test.clj

```diff
@@ -375,20 +377,158 @@
+      (is (= [1 2 3] (coerce (coerce/coll-of coerce/integer) [1 2 3])))
+      (is (= [1 2 3 nil 5] (coerce (coerce/coll-of coerce/any) [1 2 3 nil 5])))
+      (is (= [1 2 3 nil 5] (coerce (coerce/coll-of (coerce/one-of coerce/integer coerce/null)) [1 2 3 nil 5])))
+      (is (thrown? LuaError (coerce (coerce/coll-of coerce/integer) [1 2 3 nil 5])))
+      (is (= [nil nil nil nil 5] (coerce (coerce/coll-of (coerce/one-of coerce/integer coerce/null)) {5 5})))
+      (is (thrown? LuaError (coerce (coerce/coll-of coerce/integer) {5 5})))
+      (testing "min-count"
+        (is (= [1 2] (coerce (coerce/coll-of coerce/any :min-count 2) [1 2])))
+        (is (= [1 2 3] (coerce (coerce/coll-of coerce/any :min-count 2) [1 2 3])))
+        (is (thrown? LuaError (coerce (coerce/coll-of coerce/any :min-count 2) [1]))))
+      (testing "distinct"
+        (is (= [1 2 3] (coerce (coerce/coll-of coerce/any :distinct true) [1 2 3])))
+        (is (thrown? LuaError (coerce (coerce/coll-of coerce/any :distinct true) [1 2 1])))))
+
+    (testing "record"
+      (is (= {:a 1} (coerce (coerce/record :req {:a coerce/integer}) {:a 1})))
+      (is (= {"a" 1} (coerce (coerce/record :req {"a" coerce/integer}) {:a 1})))
+      (is (thrown? LuaError (coerce (coerce/record :req {:a coerce/integer}) {})))
+      (is (thrown? LuaError (coerce (coerce/record :req {:a coerce/integer}) {:a "not-an-int"})))
+      (is (thrown? LuaError (coerce (coerce/record :req {:a coerce/integer}) "not a table")))
+      (is (= {} (coerce (coerce/record :opt {:a coerce/integer}) {})))
+      (is (= {} (coerce (coerce/record :opt {:a coerce/integer}) {:x 1 :y 2})))
+      (is (thrown? LuaError (coerce (coerce/record :opt {:a coerce/integer}) {:x 1 :y 2 :a "not-an-int"})))
+      (is (= {:a 1} (coerce (coerce/record :opt {:a coerce/integer}) {:x 1 :y 2 :a 1})))
+      (is (= {:a 1} (coerce (coerce/record
+                              :req {:a coerce/integer}
+                              :opt {:b coerce/integer})
+                            {:a 1})))
+      (is (= {:a 1 :b 2} (coerce (coerce/record
+                                   :req {:a coerce/integer}
+                                   :opt {:b coerce/integer})
+                                 {:a 1 :b 2}))))
+
+    (testing "one-of"
+      (is (= "foo" (coerce (coerce/one-of coerce/string coerce/integer) "foo")))
+      (is (= 12 (coerce (coerce/one-of coerce/string coerce/integer) 12)))
+      (is (thrown? LuaError (coerce (coerce/one-of coerce/string coerce/integer) false))))
+
+    (testing "by-key"
+      (let [by-key (coerce/by-key :type
+                                  {:rect (coerce/record :req {:width coerce/number
+                                                              :height coerce/number})
+                                   :circle (coerce/record :req {:radius coerce/number})})]
+        (is (= {:type :rect :width 15 :height 20}
+               (coerce by-key {"type" "rect" "width" 15.0 "height" 20.0})))
```

Comment:
Why doesn't this fail? 🤔

`15` would not be equal to `15.0` under normal rules.

Reply:
We are checking deserialization from Lua, and Lua always represents whole numbers like `1.0` as integers. So, when we convert a Clojure map `{"type" "rect" "width" 15.0 "height" 20.0}` to Lua, it will be serialized as `{type = "rect", width = 15, height = 20}` Lua table.